rust/src/dataset/progress.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+
+use crate::format::Fragment;
+use crate::Result;
+
+/// Progress of writing a [Fragment].
+///
+/// When start writing a [`Fragment`], WriteProgress::begin() will be called before
+/// writing any data.
+///
+/// When stop writing a [`Fragment`], WriteProgress::complete() will be called after.
+#[async_trait]
```

Comment:
Might be worth being clear about which parts of the passed fragment are meaningful.
```suggestion
/// When stop writing a [`Fragment`], WriteProgress::complete() will be called after.
///
/// Note: the fragment id in the fragment passed to the `begin` and `complete`
/// methods is not necessarily meaningful. Within [Dataset::write()], for example,
/// fragment ids are not assigned until after `complete()` is called, so the passed
/// fragments will all have ids of zero.
#[async_trait]
```

Reply:
Hmm, this is problematic. As. the `.inprocess` and `fragment_{}.json` are built from fragment id. Should we use another uuid or so for these file names?

Reply:
Well the caller probably needs to be able to tie this ID back to their job, right? Otherwise they won't know which one to re-run.

Reply:
We might need to redesign the commit function as part of this, so it is compatible with conflict resolution. We can't pre-assign fragment IDs, as they need to be re-computed if there's a conflicting write we are merging with.